File: test/core/utils/bounded_concurrency_test.dart
import 'package:clashkingapp/core/utils/bounded_concurrency.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('caps work at 25 operations while preserving input order', () async {
    var active = 0;
    var peakActive = 0;

    final results = await mapWithConcurrencyLimit(
      List.generate(60, (index) => index),
      (index) async {
        active++;
        if (active > peakActive) peakActive = active;
        await Future<void>.delayed(Duration.zero);
        active--;
        return index * 2;
      },
    );

    expect(peakActive, maxConcurrentTagRequests);
    expect(results, List.generate(60, (index) => index * 2));
  });

  test('normalizes invalid concurrency limits into an argument error', () {
    expect(
      () => mapWithConcurrencyLimit(
        const [1],
        (value) => value,
        maxConcurrent: 0,
      ),
      throwsArgumentError,
    );
  });

  test('shares the default cap across simultaneous fan-outs', () async {
    var active = 0;
    var peakActive = 0;

    Future<int> operation(int value) async {
      active++;
      if (active > peakActive) peakActive = active;
      await Future<void>.delayed(const Duration(milliseconds: 1));
      active--;
      return value;
    }

    await Future.wait([
      mapWithConcurrencyLimit(List.generate(40, (index) => index), operation),
      mapWithConcurrencyLimit(List.generate(40, (index) => index), operation),
    ]);

    expect(peakActive, lessThanOrEqualTo(maxConcurrentTagRequests));
  });
}

File: expo/src/core/utils/bounded-concurrency.test.ts
import { mapWithConcurrencyLimit, MAX_CONCURRENT_TAG_REQUESTS } from './bounded-concurrency';

describe('mapWithConcurrencyLimit', () => {
  it('caps work at 25 operations and preserves input order', async () => {
    let active = 0;
    let peakActive = 0;
    const values = Array.from({ length: 60 }, (_, index) => index);
    const results = await mapWithConcurrencyLimit(values, async (value) => {
      active += 1;
      peakActive = Math.max(peakActive, active);
      await Promise.resolve();
      active -= 1;
      return value * 2;
    });

    expect(peakActive).toBe(MAX_CONCURRENT_TAG_REQUESTS);
    expect(results).toEqual(values.map((value) => value * 2));
  });

  it('rejects invalid concurrency limits', async () => {
    await expect(mapWithConcurrencyLimit([1], (value) => value, 0)).rejects.toBeInstanceOf(
      RangeError,
    );
  });

  it('shares the default cap across simultaneous fan-outs', async () => {
    let active = 0;
    let peakActive = 0;
    const operation = async (value: number) => {
      active += 1;
      peakActive = Math.max(peakActive, active);
      await new Promise<void>((resolve) => setTimeout(resolve, 1));
      active -= 1;
      return value;
    };

    await Promise.all([
      mapWithConcurrencyLimit(
        Array.from({ length: 40 }, (_, index) => index),
        operation,
      ),
      mapWithConcurrencyLimit(
        Array.from({ length: 40 }, (_, index) => index),
        operation,
      ),
    ]);
    expect(peakActive).toBeLessThanOrEqual(MAX_CONCURRENT_TAG_REQUESTS);
  });

  it('waits for queued work before rethrowing the first input-ordered error', async () => {
    const finished: number[] = [];
    await expect(
      mapWithConcurrencyLimit(
        [0, 1, 2, 3],
        async (value) => {
          await Promise.resolve();
          finished.push(value);
          if (value === 1 || value === 3) throw new Error(`failure-${value}`);
          return value;
        },
        2,
      ),
    ).rejects.toThrow('failure-1');
    expect(finished.sort()).toEqual([0, 1, 2, 3]);
  });
});
